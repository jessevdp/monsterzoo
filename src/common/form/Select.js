import { Component, renderTemplate } from '@local/system';
import './styles/Select.scss';

export default class Select extends Component {
    constructor(name, label, options = []) {
        super();
        this.setState({
            name,
            label,
            options,
            value: options[0],
        });
    }

    view() {
        const template = ''
            + '<div class="select bg-300">'
                + '<label class="text-secondary">{{label}}</label>'
                + `<select name="{{name}}"${this.isDisabled() ? ' disabled' : ''}>`
                    + '{{#options}}{{{.}}}{{/options}}'
                + '</select>'
                + '<svg class="icon" viewBox="0 0 9 7" xmlns="http://www.w3.org/2000/svg"><path d="M4.5 7L9 0H0z" fill-rule="evenodd"></path></svg>'
            + '</div>';

        const options = this.state.options.map(option => {
            const selected = this.state.value === option ? ' selected' : '';
            return `<option value="${option}"${selected}>${option}</option>`;
        });

        return renderTemplate(template, {...this.state, options});
    }

    events() {
        this.on('change', 'select', () => {
            const $select = this.getHTMLElement('select');
            const isActive = document.activeElement === $select;
            this.setState({ value: $select.value });
            if (isActive) this.getHTMLElement('select').focus();
        });
    }

    isDisabled() {
        return this.state.options.length < 2;
    }


    set value(value) {
        this.setState(state => {
            if (state.options.includes(value)) return { value };
            else throw new RangeError(); 
        });
    }

    get value() {
        return this.state.value;
    }

    set options(options) {
        this.setState(state => {
            const value = options.includes(state.value) ? state.value : options[0];
            return { options, value };
        });
    }

    get options() {
        return this.state.options;
    }
}