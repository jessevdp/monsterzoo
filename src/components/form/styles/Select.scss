@import '~@local/styles/variables/spacing';
@import '~@local/styles/mixins/form-reset';

.select {
    display: flex;
    flex-direction: column;
    padding: spacing(1) spacing(2) spacing(2);
    position: relative;

    > label {
        font-size: .6rem;
        letter-spacing: .1em;
        text-transform: uppercase;
    }

    > select {
        @include select-reset;
        font-size: 1rem;
        margin-top: spacing(1);
        margin-left: spacing(2);
        text-transform: capitalize;
    }

    > .icon {
        height: .5rem;
        position: absolute;
        top: 50%;
        right: spacing(3);
        transform: translateY(-50%);
        > path {
            fill: var(--ui-contrast-tertiary);
        }
    }

    > select:disabled ~ .icon {
        display: none;
    }
}