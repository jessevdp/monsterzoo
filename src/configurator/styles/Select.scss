@import '~@local/styles/variables/spacing';
@import '~@local/styles/mixins/form-reset';

.select {
    display: flex;
    flex-direction: column;
    padding: spacing(1) spacing(2) spacing(2);

    > label {
        font-size: .6rem;
        letter-spacing: .1em;
        text-transform: uppercase;
    }

    select {
        @include select-reset;
        font-size: 1rem;
        margin-top: spacing(1);
        margin-left: spacing(2);
        text-transform: capitalize;
    }
}