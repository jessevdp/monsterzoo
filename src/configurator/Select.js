import { Component, renderTemplate } from '@local/system';
import './styles/Select.scss';

export default class Select extends Component {
    constructor(name, label, options) {
        super();
        this.setState({
            name,
            label,
            options,
            value: options[0],
        });
    }

    view() {
        const template = ''
            + '<div class="select bg-300">'
                + '<label class="text-secondary">{{label}}</label>'
                + '<select name="{{name}}">'
                    + '{{#options}}{{{.}}}{{/options}}'
                + '</select>'
            + '</div>';

        const options = this.state.options.map(option => {
            const selected = this.state.value === option ? ' selected' : '';
            return `<option value="${option}"${selected}>${option}</option>`;
        });

        return renderTemplate(template, {...this.state, options});
    }

    events() {
        this.on('change', 'select', () => {
            const $select = this.getHTMLElement('select');
            const isActive = document.activeElement === $select;
            this.setState({ value: $select.value });
            if (isActive) this.getHTMLElement('select').focus();
        });
    }


    set value(value) {
        this.setState({ value });
    }

    get value() {
        return this.state.value;
    }

    set options(options) {
        this.setState({ options });
    }

    get options() {
        return this.state.options;
    }
}