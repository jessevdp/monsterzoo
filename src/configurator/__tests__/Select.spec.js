import Select from '../Select';

it('renders correctly', () => {
    const options = ['foo', 'bar'];
    const select = new Select('name', 'label', options);
    expect(select.view()).toMatchSnapshot();
    select.cleanup();
})

describe('constructor', () => {
    it('sets the [name] on state', () => {
        const name = 'foobar';
        const select = new Select(name, 'label');
        expect(select.state.name).toBe(name);
        select.cleanup();
    })
    it('sets the [label] on state', () => {
        const label = 'label';
        const select = new Select('name', label);
        expect(select.state.label).toBe(label);
        select.cleanup();
    })
    it('sets the [options] on state', () => {
        const options = ['foo', 'bar'];
        const select = new Select('name', 'label', options);
        expect(select.state.options).toBe(options);
        select.cleanup();
    })
})

describe('[options] setter', () => {
    it('updates the [options] on state', () => {
        const newOptions = ['foo', 'bar'];
        const select = new Select('name', 'label', ['foo']);
        select.options = newOptions;
        expect(select.state.options).toBe(newOptions);
        select.cleanup();
    })
    it ('uses the setState method', () => {
        const newOptions = ['foo', 'bar'];
        const select = new Select('name', 'label', ['foo']);
        const spy = jest.spyOn(Select.prototype, 'setState');
        select.options = newOptions;
        expect(spy).toHaveBeenCalledWith({ options: newOptions });
        select.cleanup();
    })
})

describe('[options] getter', () => {
    it('returns the value of state.options', () => {
        const options = ['foo', 'bar'];
        const select = new Select('name', 'label', options);
        expect(select.options).toBe(options);
        select.cleanup();
    })
})