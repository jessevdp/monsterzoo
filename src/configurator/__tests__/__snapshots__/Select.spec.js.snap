// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `"<div class=\\"select bg-300\\"><label class=\\"text-secondary\\">label</label><select name=\\"name\\"><option value=\\"foo\\">foo</option><option value=\\"bar\\">bar</option></select></div>"`;
